### ISSUE TITLE
AA bluetooth Phone call immediately disconnect post dial or post pick up.

### ISSUE THREAD

**Issue by Jinmcx** [gartnera/headunit#2]
AA bluetooth Phone call immediately transfer to phone post dial or post pick up. When receiving phone calls via bluetooth the MZD screen will pop up and immediately transfer to phone. The only fix right now is kill AA or reboot (without AA open) to pick up calls. I'm pretty sure it's similar to dialing. I assume a fix would be related to having something conflicting with AA and MZD to disconnect right after a call is dial/pick up.

Similar to what ariehm said below, i have the option to transfer to headset but it just instantly transfers back to phone no matter how many times I hit transfer.

Solution: I have no clue :(

**Comment by ariehm** [comment]
@Jinmcx are you sure that the call actually disconnects? i experience a similar issue when making or receiving calls while in Android Auto, except rather than disconnecting, the call audio is transferred to the handset and does not utilize the car's microphone or speakers.

can you please try the following, to make sure that your and my issues are the same:

1. put your car into "accessory" mode
2. once the infotainment system has booted fully, start Android Auto
3. make (or receive) a phonecall
4. observe that the Android Auto ui drops away to the Mazda Connect phonecall ui
5. place the phone to your ear

is the call still in progress and is the audio coming out of the handset's earpiece? in my experience, this is the case. i have also tried tapping the "transfer audio to car" button in the Mazda Connect phonecall screen, and it does so briefly before sending it back to the handset again.

**Comment by Jinmcx** [comment]
You're completely right, you describe just what I experience. I'm using the wrong terminology.

**Comment by Kennair** [comment]
Here is the result of my testing of AA in the Mazda 3:

If you have the Mazda system set to auto receive phonecalls AA will drop away when a call is received or initiated and the Mazda Connect phonecall ui pops up on the CMU, however the call is transferred to the phone with no way to re-establish it to the CMU.  In fact attempting to transfer back to the CMU results in a loop of the system trying to grab the call then dropping back to the phone.  All the while the caller remains on the cell phone.

I then tested by deselecting Auto Answer phonecalls in the Mazda system which resulted in the CMU no longer dropping to the Mazda Connect phoncall ui, instead AA now showed the phonecall in progress screen BUT it had a message saying Bluetooth Disconnected, which probably accounts for the call transferring to the phone, however Bluetooth wasn't actually disconnected as I could still hear music from Spotify that was playing via the phone and bluetooth to the car.

Hope that helps.

Ken.

**Comment by borconi** [comment]
This might sound stupid, but you might want to try it, in hu_aap.c comment out line 425:
`   if (file_get ("/tmp/mnt/sdnav/hu_disable_audio_out"))    	`
or indeed as the if suggest do a:
`touch /tmp/mnt/sdnav/hu_disable_audio_out`
to create a file which will disable the audio.

That will prevent the audio being synced from the phone to the headunit over USB, now on your phone make sure you use bluetooth audio for both phone calls and media and see if the problem persist.

Basically you're not transferring any audio over the USB all audio will go trough bluetooth music (A2DP) and phone as well.

**Comment by Jinmcx** [comment]
This sounds like the proper solution :] I wonder if mic would go through, don't have the time to test it ><.

**Comment by borconi** [comment]
Sorry I don't have a Mazda to test it :) Yes mic should go trough

**Comment by Jinmcx** [comment]
maybe we can wait for mr gartnera :[, Idk how to edit the files. Last time I told them i'm on a mac and how do I install terminal they laugh ></

**Comment by Jinmcx** [comment]
Creating a /tmp/mnt/sdnav/hu_disable_audio_out file disables usb audio but it doesn't resolve the CMU talk interface from kicking the phone call back to the phone instead of the cars speaker/mic. The created file/folder gets delete upon reset of the CMU but it test-able during same instance the file is created.

**Comment by youssefmbazan** [comment]
Is there any solution for this issue? its very frustrating :/

**Comment by Jinmcx** [comment]
None yet aside from killing AA, here's the 2 action plan according gartnera (taken from chat).

1) find the correct audio id every boot. Right now it gets the wrong one sometimes and you can't hear any audio
2) report to AA on the phone that we are connected via bluetooth. When the call comes in, it will get routed through Bluetooth. But we need to figure out how to report that we have bluetooth in headunit

**Comment by Teecoder300** [comment]
The way I "fixed" it is by just deselect Mazda call option in my Bluetooth
options. Just have to put the phone on speaker and listen there. Yea it's
not as loud but at least you can still have a phone conversation.

**Comment by Jinmcx** [comment]
I found a work around with the phone and AA. Simply be in call (call 611) prior to turning on AA. Afterwards all phone seem to pick up and calls from the same instance of AA without that disconnect back to phone. :) This is working on .93 tested it for 2 days and still working. Dialing will still need to be done in mazda interface but AA seems to be always running in the background.

**Comment by Teecoder300** [comment]
Could you explain it again. I read it twice and still don't get the steps.
1. Make a call prior to turning on Android Auto App. (Is your USB already
connected?
2. Connect to AA.
3. Hang up the call?
4. All incoming or outgoing calls no longer switch back and forth?

**Comment by Jinmcx** [comment]
@Teecoder300 
My setup is different from yours but the steps should work the same. I use a AA phone that's always in the car and connected. I use a calling phone (in my pocket) separate.

Let's say you need to make a call you/expecting a call. You would make a phone call while the device is connected via cmu bluetooth and cable (without running AA, calling 611). During the phone call press home button (on the command knob) and open up AA during the call. During this instances of AA if you make a phone call or receive call (via cmu) you can answer and it won't disconnect you back to your phone. You can always go back to AA during the phone call as well navigation home -> Application -> android (cause it's always in the background.) 

Does that make sense, I might make a video if it seems confusing.

**Comment by Teecoder300** [comment]
I think I understand it from there. Thank you. When I try it and it doesn't
work I'll maybe request a video gaga

**Comment by Jinmcx** [comment]
Here's a quick video demo and it uses the typically AA setup (using a single phone for both bluetooth and android auto). It's a little different when you use the alternative setup (using one phone for AA and one phone for bluetooth). 

https://www.youtube.com/watch?v=RrLtrDK9edY

Instructions: (Similar to video above)
Make Call -> Open up AA -> (disconnected phone or whatever) -> call again via AA or CMU -> CMU menu call pops up answer and phone doesn't connect. -> Disconnect via cmu or button etc.
Some A7X in the background =]!

Note minor but you can always exit out of AA (not via home but by the AA exit menu) and make call and turn on AA to get into this phone ready state.

**Comment by Teecoder300** [comment]
Jinmcx. Your way seems to have worked. Awesome thank you

**Comment by youssefmbazan** [comment]
I dont seem to understand what Jinmcx is trying to say, can someone please re-explain?

**Comment by Teecoder300** [comment]
There is a link to a YouTube video he message. Go to that. Watch it and
things will be clear.

**Comment by davepro305** [comment]
How is work progress going on this issue? Any plan to fix it in the next release? (Would be awesome!)

**Comment by gartnera** [comment]
I've put about 20 hours into a fix, but I was unable to make any progress. If I had a real headunit, I might be able to make some progress, but unfortunately I don't.

The problem is the headunit application should be reporting to the phone that bluetooth is connected to the car, but it isn't doing that. 

The other way to figure out what is going on is some reverse engineering of the android applications. However, while my C reverse engineering skills are decent, but I can't make heads or tails of the Android Application. There are a lot of moving components (the "userspace" android application, google play services, system stuff) that I have a hard time keeping track of.

Also I'm using an iPhone now, so I lack the motivation to really dig much deeper.

**Comment by borconi** [comment]
Please remove your last sentence, that is like swearing to an Android community 😄

**Comment by davepro305** [comment]
You can come home and try on my mazda if you wish :p loool

**Comment by goon818** [comment]
Can we try to switch the audio from USB-Audio to Bluetooth.
I think, our problem is that we use USB-Audio.

**Comment by borconi** [comment]
Alex, I know you are an Apple desertor, but I think I just got one tiny step closer to solving the mystery....
Mike, when he reverse engineered the app he defined 7 channels, but there are more than that.., For example one of the channels is the BluetoothChannel, and one is the PhoneAdapter channel. To get this working correctly those channels setup and data needs to be cracked.

I have managed to figure out the value of the BluetoothChannel but that is only the first step because, you also need to encode what kind of pairing it supports and the bluetooth address as well. The second part is the really ugly one.

Just to give you some direction on it the def_buf is the key, adding: 0x0A, 6, 0x08, 8, 0x32 means bluetoothchannel, but without the rest of the array it just causes a Google Play crash, as said baby step.

**Comment by borconi** [comment]
A bit of completion:

Adding: ```0x0A, 7,   0x08,     8,      0x32, 3, 0x0A, 0x01, 0x50, ``` to the sd_buf results in the app starting correctly and Bluetooth Adapter is recognized, with address = P, the 0x50=P I wasn't able to work out how to get more letters in the string, I guess we need to somehow pass the whole HEX MAC of the bluetooth adapter to AA. Looking at the logs AA tries to connect to "P" and of course fails..... I think if I (we) can figure this out we can have it sorted.

**Comment by borconi** [comment]
Ok I made some more progress, the BluetoothManager should look like this:
```
0x0A, 26,   0x08,     8,      0x32, 22, 0x0A, 17, 'F','C',':','5','8',':','F','A',':','1','2',':','1','A',':','0','D',0x12,1,0x01,
```
I'm not sure if we need or not the ':' in the bluetooth address, the tail part is the bluetooth connection mode, and I need to figure out what number means what, current string means mode 1, but I don't know what AA means by mode one... I guess it has to do something with bluetooth profiles need so read the Android documentation to see if they give any constants for A2DP, HFP, etc. Goes without saying that the above string is my bluetooth device MAC you will need to update with the headunit CMU (each one is different!)

Hope this gives you the needed motivation to get on cracking with it!

**Comment by borconi** [comment]
Ok I can confirm that I got it working in my tests, here it what it needs changing:

hu_aap.c file, the sd_buf variable needs the following string:
```
0x0A, 27,   0x08,     8,      0x32, 23, 0x0A, 17, 'F','C',':','5','8',':','F','A',':','1','2',':','1','A',':','0','D',0x12,2,0x02,0x03,
```
Of course replace the MAC address with your own bluetooth MAC address. It does need the full address with semicolons as well. mode 0x02 means A2DP support (at the end of the string). 

In the same file the  aa_type_ptr_t aa_type_array [AA_CH_MAX + 1] [3] [32] = { needs to be extended with:
```
// Channel 8 Bluetooth Manager:
    NULL, NULL, NULL, NULL, NULL, NULL, NULL, aa_pro_all_a07, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
    NULL, aa_pro_mic_b01, NULL, NULL, aa_pro_mic_b04, aa_pro_mic_b05, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
```

hu_aap.h also needs a bit of tweaking:
from:
```
 #define AA_CH_MAX 7
```
to:
```
 #define AA_CH_MAX 8
```

You should be all good to go now, you only need to figure it out how to get the MAC address of the bluetooth adapter and pass it on to the hu_app.c
The phone will now report as being connected instead of disconnected and it means it will work correctly even with Android Auto 2.0 without the need of switching to developer mode.

**Comment by izacus** [comment]
Worth testing out just to see what the phone does in that configuration - even if the call gets handed to CMU it's still significantly better than having it distonnect from the car completely.

**Comment by borconi** [comment]
Yes just I don't have a Mazda.... so I cannot do anything more I'm affraid somebody will need to implement the changes and make a build, and test it. 
However I tested this changes on my Android version of the app, and I can tell you that now when you connect the phone if Bluetooth is not switched on, AA will automatically switch it on for you and initiate the pairing connection. Also this fixed the missing Dialpad in AA 2.0, so there is no more need for enabling the developer mode.

**Comment by Jinmcx** [comment]
Let me know if you want me to test anything out :] I haven't had any AA issue or problems yet with my Yaris iA.

**Comment by anod** [comment]
Bluetooth service proto definiton:
```
    message BluetoothService {
        enum BluetoothPairingMethod
        {
            BLUETOOTH_PARING_METHOD_1 = 1;
            BLUETOOTH_PARING_METHOD_2 = 2;
            BLUETOOTH_PARING_METHOD_3 = 3;
            BLUETOOTH_PARING_METHOD_4 = 4;
        }
        required string car_address = 1;
        repeated BluetoothPairingMethod supported_pairing_methods = 2;
    }
```

Not sure what are paring methods, but followed @borconi message and the following initialization works for me, shows the phone tab again:

```
        Service bluetooth = new Service();
        bluetooth.id = 8;
        bluetooth.bluetoothService = new Service.BluetoothService();
        bluetooth.bluetoothService.carAddress = <btMacAddress>;
        bluetooth.bluetoothService.supportedPairingMethods = new int[] { 2, 3 };
```

**Comment by borconi** [comment]
@anod PairingMode = 2 and 4 means A2DP+HFP profile, other modes are reported as unsupported so not sure either what are the other modes...

**Comment by izacus** [comment]
Hmm, I've tested the patch and indeed the call screen now appears. However, we still get kicked outside of AA and the call gets redirected back to phone. I wonder if we should NOT signal HFP capability and let phone handle everything.

**Comment by goon818** [comment]
Can You release the patch. And what have you changed?

**Comment by davepro305** [comment]
I'm fine with the calls made with the mazda stock app. As long as they work

**Comment by goon818** [comment]
Forget the USB audio. In the HUR from Borconi the telephone calls over usb doesn't too. Read the information text from him.

**Comment by borconi** [comment]
@izacus that was the case with Android Auto 1.6 and with 2.0 it's the same if you do not include the patch. Now I'm not sure why this is happening on the mazda, without signalling  HFP and letting the phone do the work, like you said, it works just fine with my handsfree, so it's really interesting on why the Mazda CMU has this behaviour. Can you try de-selecting the A2DP profile and the Contact sharing from your Bluetooth menu? Maybe even forget and re-set up the pairing. It's something awkward with the CMU starting the handsfree. Or maybe the CMU has some problem with the audio focus??? Just thinking, is it possible that the CMU handsfree app is not able to gain audio focus due to Android Auto running in the background and using the audio channel, which will force the CMU handsfree app to hand back the call to the phone? 

Just thinking....

**Comment by izacus** [comment]
I didn't have time to look into it. But what's probably happening from things I've seen:

1. Call comes in and CMU intercepts it
2. Because CMU shows it's own UI, we get kicked out of headunit
3. Because we signaled that audio/video channels are available during calls, the phone probably wants to route audio/mic over headunit
4. Because headunit died due to CMU interfering, the call gets stuck on the phone.

Hence my question about setting those channels as not available.

**Comment by borconi** [comment]
Regardless of what you signal from the headunit, the phone audio is ALWAYS
rooted through bluetooth for the phone calls. This is how all the official
headunits works as well. If you wish you can signal VIDEO_FOCUS and
AUDIO_FOCUS lost to AA when headunit it's kicked out, that assuming you
have something similar to Android ```onDestroy``` or ```onPause```, I'm not
sure how the lifecycle of the C app looks on the CMU.



Kind regards
Emil Borconi-Szedressy
Tel: +44-791-32-32-588

**Comment by borconi** [comment]
I ain't linux and C expert, neither know how to code properly,  but I think your problem is not as much as from AA on the phone, but from the app on the headunit. I know the code below is not correct but the purpose is to give you an indication/hint/idea on what might be the cause of problem. 

@izacus @lmagder maybe you can write the code correctly and see if that helps fixing the issue.

I think that this 3 lines in the main.c are somehow blocking the CMU phone app of using the Audio channels properly:
```
mic_pipeline = gst_parse_launch("alsasrc name=micsrc ! audioconvert ! audio/x-raw-int, signed=true, endianness=1234, depth=16, width=16, channels=1, rate=16000 ! queue !appsink name=micsink async=false emit-signals=true blocksize=8192",&error);

aud_pipeline = gst_parse_launch("appsrc name=audsrc is-live=true block=false max-latency=1000000 ! audio/x-raw-int, signed=true, endianness=1234, depth=16, width=16, rate=48000, channels=2 ! volume volume=0.4 ! alsasink ",&error);

au1_pipeline = gst_parse_launch("appsrc name=au1src is-live=true block=false max-latency=1000000 ! audio/x-raw-int, signed=true, endianness=1234, depth=16, width=16, rate=16000, channels=1 ! volume volume=0.4 ! alsasink ",&error);

```

I think you should modify hu_app.c to have something similar to this:

```
 int aa_pro_aud_b02 (int chan, byte * buf, int len) {                  // 08-22 20:03:09.075 D/ .... hex_dump(30767): S 4 AUD b 00000000 08 00 10 01   Only at stop ??
    if (len != 2)//4 || buf [2] != 0x08)
      loge ("Audio Sink Stop Request");
    else
      logd ("Audio Sink Stop Request");//: %d", buf [3]);
    if (chan == AA_CH_AUD)
      out_state_aud = 1;                                                      // Signal Audio stop
    else if (chan == AA_CH_AU1)
      out_state_au1 = 1;                                                      // Signal Audio1 stop
    else if (chan == AA_CH_AU2)
      out_state_au2 = 1;                                                      // Signal Audio2 stop
    //hex_dump ("AOSSR: ", 16, buf, len);
	gst_object_unref(mic_pipeline);            //Destroy the Audio and Mic channel so CMU can regain control.
	gst_object_unref(aud_pipeline);
	gst_object_unref(au1_pipeline);
    return (0);
  }
```

**Comment by lmagder** [comment]
I think you guys are on the right track, but there is an extra part: the CMU had it's own audio focus system which is controlled by dbus. The code which sends the messages is in the input filter code, it doesn't have to be though, I don't think. So the correct thing is to stop the streams like @borconi suggests, but you also need to switch to the Bluetooth audio channel.

I haven't tested any of this though. Haven't had a chance to load stuff on my car and mess around for a bit

**Comment by lmagder** [comment]
Also you don't want to close the pipe lines and realease them I don't think. There is a stopped state you can set so they do nothing  but can be started again later. The microphone code does this.

**Comment by borconi** [comment]
Well it was more of an idea.... not sure who the whole CMU is working, and I don't have one so there is no way I can twinkle around to get to the bottom of it.

**Comment by izacus** [comment]
I guess we need to dump CMU  dbus events while it's handling calls.

**Comment by goon818** [comment]
Can you not get the information from the cellphone? From the CMU... This is a good idea but very heavy to finde a issue.
But i have a wish.
Can you release a version without this last issue with the cmu.
I think, i can switch back manual to AA after a call. But at time i can not make a call with AA.
This is a good idea from you, but i think that is important to release a version with a chance to make a call.

**Comment by Jinmcx** [comment]
@borconi I just have this crazy idea. I remember when we log every command with logcat can't we log to see what occurs during a phone call prior to activating android audo and compare what occurs during a phone call and activating android auto? Maybe we could create a script that runs a similar activity (making a call) and activating android auto? Just know some thoughts

**Comment by borconi** [comment]
Hi @Jinmcx this issue is not really Android Auto one, but something to do with the audio on the headunit. When the app start using the audio somehow it affects the bluetooth sink of the unit. Having a think about this @izacus , @Imagder I have a workaround idea for this.

It goes as follows:
1) Only declare the SYSTEM audio channel in the protobuf (you need to have at least 1 media channel declared otherwise AA won't start on the phone throwing error 2)
2) do not start/use any of the mic/audio pipeline, which will mean that no audio is outputted through the app
3) The phone will connect the the A2DP when it connects to the car, and since we only declared the system channel in the protobuf only the system notification will be sink to the app everything else will stay on the phone, but since the phone is connected to an A2DP sink it will playback everything trough that.

Now this was they headunit app doesn't interfere with the CMU audio at all and when you have an incoming call everything should work normally.

Just a thought...

**Comment by lmagder** [comment]
That might work, but we are also breaking lots of other non-phone call stuff 😄 I would prefer to fix the underlying issue. With @izacus most revent PR integrated I can place a phonecall through the AA UI and in fact the car audio switches to BT but we aren't telling the phone that so things get weird and it transfers the call the phone's speakerphone. Also the UI on the car is disabled since the app has stolen input focus, but the CMU has changed the surface back to opera to display the call UI. This tells me we are really close to getting this working.

**Comment by borconi** [comment]
Proper fix is always better. Off topic question, do you guys (@izacus, @lmagder ) have experience with Ofono/PulseAudio/Bluez? And have you ever worked with a C.H.I.P. SBC (https://getchip.com/pages/chip) I'm trying to build a small handsfree from this device, A2DP works smoothly, HFP profile does come up and phone connects to it but when making a call there is no sound at all (not on the phone, not on the device) at 2nd-3rd attempt of call the PulseAudio on the device just dies with a Segfault so I'm not sure where to start looking for the problem (not a linux guru myself)

**Comment by lmagder** [comment]
@borconi Sorry don't have much experience. I poked around at the Ofono D-Bus interface through D-Feet and was able to place a call on my phone, get status etc, but that's about it. I didn't actually try if the audio works since I didn't wait for the person to pickup. Tested A2DP, which worked though. This is with a Broadcom BT USB with stock drivers/FW on Ubuntu 16.10 x64

My plan was to use the Ofono D-Bus API to do the HFP for the desktop linux version of the code but I haven't actually tried yet. My thinking would that it would be easier to debug the phone-related protocol stuff this way than on the car.

**Comment by borconi** [comment]
Thanks @lmagder. No this is a bigger think I want to do, working on an Android based headunit (which will be powered by a Odroid C2), but Android doesn't support A2DP and HFP profile. So inside the unit there will be a secondary board (cheaper better) which will have only one function, run A2DP and HFP. The secondary board will be connected to the Odroid, and I will need to create an app which sits on the Odroid, looks like a phone dialer, but actually all it does it's communicating with the secondary board... 😄 simple, ha? I had success with a Raspberry Pi 2, I was able to use it as a fully working bluetooth headset, but the RPI2 it's quite expensive for a bluetooth module. The chip board costs only $9!!!! and it has Wifi+Audio+Bluetooth on board + 8gb EMMC so it will be ideal for the job, but I'm struggling to get the audio working... Ok, I think it's enough with the hi-jack of the thread 😃

**Comment by Jinmcx** [comment]
@borconi lame was gonna ask if you guys were excited for ryzen :'(

**Comment by goon818** [comment]
If i can, I would take the raspberry pi in my car.
I have found none solution for it.

**Comment by borconi** [comment]
Raspberry Pi 2, works fine as a bluetooth headtset, I'm having trouble with
Raspberry Pi Zero.


Kind regards
Emil Borconi-Szedressy
Tel: +44-791-32-32-588

**Comment by EQMOD** [comment]
the following headunit logs shows up during the call - it seems it toggles between USB and BTHF. Any idea guys ?

AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"USB","streamModeName":"USB","newFocus":"temporarilyLost","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 13 for stream USB
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"BTHF","streamModeName":"BTHF","newFocus":"gained","destination":"Cabin","focusType":"transient","streamType":"Telephony"}
Found audio sessionId 15 for stream BTHF
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"BTHF","streamModeName":"BTHF","newFocus":"lost","destination":"Cabin","focusType":"transient","streamType":"Telephony"}
Found audio sessionId 15 for stream BTHF
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"USB","streamModeName":"USB","newFocus":"gained","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 13 for stream USB
E: /home/songrit/repo/gartnera-master/hu/hu_aap.cpp:922: iaap_msg_process : Unknown chan: 8
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"USB","streamModeName":"USB","newFocus":"temporarilyLost","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 13 for stream USB
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"BTHF","streamModeName":"BTHF","newFocus":"gained","destination":"Cabin","focusType":"transient","streamType":"Telephony"}
Found audio sessionId 15 for stream BTHF
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"BTHF","streamModeName":"BTHF","newFocus":"lost","destination":"Cabin","focusType":"transient","streamType":"Telephony"}
Found audio sessionId 15 for stream BTHF
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"USB","streamModeName":"USB","newFocus":"gained","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 13 for stream USB
E: gps/mzd_gps.cpp:64: process_gps : GPS parse error: GPS Data Bad Format [$GPVTG] :: Checksum is invalid!.
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"USB","streamModeName":"USB","newFocus":"temporarilyLost","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 13 for stream USB
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"BTHF","streamModeName":"BTHF","newFocus":"gained","destination":"Cabin","focusType":"transient","streamType":"Telephony"}
Found audio sessionId 15 for stream BTHF
E: gps/mzd_gps.cpp:64: process_gps : GPS parse error: GPS Data Bad Format [$GPGSV] :: Checksum is invalid!.
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"BTHF","streamModeName":"BTHF","newFocus":"lost","destination":"Cabin","focusType":"transient","streamType":"Telephony"}
Found audio sessionId 15 for stream BTHF
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"USB","streamModeName":"USB","newFocus":"gained","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 13 for stream USB

**Comment by borconi** [comment]
Hi.

Might not be relevant, but I cam across a similar problem on Honda head units.  One of my usesrs installed the app on a Honda 2015 which came without AA support. 
On the original version he had the app crashing at every phone call, the fix was to stop/release all audio in the app when the phone call was coming in, then recreate the streams, I still think that this is somehow related to your issue as well, the CMU bluetooth app cannot get a lock/hold of the audio streams because they are being in used by the AA, so if you haven't tried yet, try fully releasing/offloading any audio when there is a phone call.

**Comment by EQMOD** [comment]
here some info - at AA headunit start, even if there is no active audio, AA starts to set audio focus to USB


W: /home/songrit/repo/gartnera-master/hu/hu_aap.cpp:631: hu_handle_AudioFocusRequest : AudioFocusRequest Focus Request CTR: 2
requestAudioFocus({"sessionId":13})
{}
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"AUX","streamModeName":"AUX","newFocus":"lost","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 1 for stream AUX
AudioManagerClient::Notify signalName=audioFocusChangeEvent payload={"streamName":"USB","streamModeName":"USB","newFocus":"gained","destination":"Cabin","focusType":"permanent","streamType":"Media"}
Found audio sessionId 13 for stream USB
W: /home/songrit/repo/gartnera-master/hu/hu_aap.cpp:631: hu_handle_AudioFocusRequest : AudioFocusRequest Focus Request CTR: 4

**Comment by hansaya** [comment]
I have the same issue. Mine does not even work with AA not in use. Only time I can take calls is at bootup. After the radio booted completely I cant take any calls. If you guys need more information I'm happy to get into the unit using ssh.

**Comment by Trevelopment** [comment]
How many times has this same issue been opened in this repo? After merging #78 you get to close like 5 issues that address the same thing.

**Comment by lmagder** [comment]
Mostly fixed
